File: IbolyaGergo-cs50-problems-2022-x-sentimental-credit/credit.py
# TODO
from cs50 import get_string

# Making sure the input is numeric
while True:
    number = get_string("Number: ")
    if number.isnumeric():
        break

# Luhn's Algorithm to check if the credit card number is valid


def checksum(number):

    revnumber = number[::-1]
    mult = ""
    for digit in revnumber[1::2]:
        mult += str(2 * int(digit))

    not_mult = revnumber[::2]

    sum = 0
    for digit in (mult + not_mult):
        sum += int(digit)
    print(sum)
    if str(sum)[-1] == "0":
        return True
    else:
        return False


if checksum(number):
    if len(number) == 15 and int(number[0]) == 3 and int(number[1]) in (4, 7):
        card = "AMEX"
    elif len(number) == 16 and int(number[0]) == 5 and int(number[1]) in range(1, 6):
        card = "MASTERCARD"
    elif (len(number) == 13 or len(number) == 16) and int(number[0]) == 4:
        card = "VISA"
    else:
        card = "INVALID"
else:
    card = "INVALID"
print(card)

File: IbolyaGergo-cs50-problems-2019-x-credit/credit.c
#include <stdio.h>
#include <cs50.h>

    
int sum=0;
int counter=1;
int dig, sec, dignum, i;

int main(void)
{
    long card=get_long("Number: ");
    long number=card;
 
    while(counter<19)
    {
        if((number%10)>0)
        {
            dig=number%10;                 /*non-zero digit counted from right to left*/
        
            if(counter%2==0&&dig>0)        /*every 2nd digit multiplied by 2*/
            {
                sec=dig*2;
                sec=sec/10+(sec%10);
                sum=sum+sec;               /*sum of digits is added to the sum*/
            }
            else
            {
                sum=sum+dig;
            }
        }
        
        number/=10;                        /*reducing the number of digits from the right*/
        
        counter++;
        
        if(number>0)
        {
            dignum=counter;
        }
        
    }
   
    if(sum%10==0)                          /*control of the validity of the card number
                                            * if valid, check if it's VISA, MasterCard or AmericanExpress*/
    {
        if(dig==4)
        {
            if(dignum==13||dignum==16)
            {
                printf("VISA\n");
            }
        }
    
        if(dignum==15)
        {
            for(i=0;i<13;i++)
            {
                card/=10;
            }

            if(card==34||card==37)
            {
                printf("AMEX\n");
            }
        }

        if(dignum==16&&dig==5)
        {
            for(i=0;i<15;i++)
            {
                card/=10;
            }
            
            if(card==1||card==2||card==3||card==4||card==5)
            {
                printf("MASTERCARD\n");
            }
        }
    }
            
    else
    {
        printf("INVALID\n");
    }
}
